pil-analyzer/src/condenser.rs

```diff
@@ -162,77 +161,31 @@ pub struct Condenser<'a, T> {
     symbol_values: BTreeMap<SymbolCacheKey, Arc<Value<'a, T>>>,
     /// Current namespace (for names of generated witnesses).
     namespace: AbsoluteSymbolPath,
-    next_witness_id: u64,
+    /// ID dispensers.
+    counters: Counters,
     /// The generated witness columns since the last extraction.
     new_witnesses: Vec<Symbol>,
     /// The names of all new witness columns ever generated, to avoid duplicates.
     all_new_witness_names: HashSet<String>,
-    new_constraints: Vec<IdentityWithoutID<AlgebraicExpression<T>>>,
-}
-
-#[derive(Debug, PartialEq, Eq, Clone)]
-pub struct IdentityWithoutID<Expr> {
-    pub kind: IdentityKind,
-    pub source: SourceRef,
-    /// For a simple polynomial identity, the selector contains
-    /// the actual expression (see expression_for_poly_id).
-    pub left: SelectedExpressions<Expr>,
-    pub right: SelectedExpressions<Expr>,
-}
-
-impl<Expr> IdentityWithoutID<Expr> {
-    /// Constructs an Identity from a polynomial identity (expression assumed to be identical zero).
-    pub fn from_polynomial_identity(source: SourceRef, identity: Expr) -> Self {
-        Self {
-            kind: IdentityKind::Polynomial,
-            source,
-            left: SelectedExpressions {
-                selector: Some(identity),
-                expressions: vec![],
-            },
-            right: Default::default(),
-        }
-    }
-
-    pub fn into_identity(self, id: u64) -> Identity<SelectedExpressions<Expr>> {
-        Identity {
-            id,
-            kind: self.kind,
-            source: self.source,
-            left: self.left,
-            right: self.right,
-        }
-    }
+    new_constraints: Vec<AnalyzedIdentity<T>>,
 }
 
 impl<'a, T: FieldElement> Condenser<'a, T> {
     pub fn new(symbols: &'a HashMap<String, (Symbol, Option<FunctionValueDefinition>)>) -> Self {
-        let next_witness_id = symbols
-            .values()
-            .filter_map(|(sym, _)| match sym.kind {
-                SymbolKind::Poly(PolynomialType::Committed) => {
-                    Some(sym.id + sym.length.unwrap_or(1))
-                }
-                _ => None,
-            })
-            .max()
-            .unwrap_or_default();
+        let counters = Counters::with_existing(symbols.values().map(|(sym, _)| sym), None, None);
```

Comment:
use `.keys()` instead of `.values().map(|(sym, _)| sym)`

Reply:
Unfortunately, not. The values are pairs. It would have worked to replace `items().map(...`, but not `values().map(...`

Reply:
Ah, sorry. I may be still sleepy.